File: week10/19/P4.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int* seats;
    int capacity;
    int size;
    int n;
} ExamRoom;

ExamRoom* examRoomCreate(int n) {
    ExamRoom* obj = (ExamRoom*)malloc(sizeof(ExamRoom));
    obj->seats = (int*)malloc(sizeof(int) * n); // 최대 n개의 좌석
    obj->capacity = n;
    obj->size = 0;
    obj->n = n - 1;
    return obj;
}

// 이진 탐색을 통한 좌석 위치 찾기
int findInsertPosition(ExamRoom* obj, int value) {
    int left = 0, right = obj->size - 1;
    while (left <= right) {
        int mid = left + (right - left) / 2;
        if (obj->seats[mid] == value)
            return mid;
        else if (obj->seats[mid] < value)
            left = mid + 1;
        else
            right = mid - 1;
    }
    return left;
}

// 좌석을 추가하는 함수
void insertSeat(ExamRoom* obj, int value) {
    int pos = findInsertPosition(obj, value);
    for (int i = obj->size; i > pos; --i) {
        obj->seats[i] = obj->seats[i - 1];
    }
    obj->seats[pos] = value;
    obj->size++;
}

// 좌석을 삭제하는 함수
void removeSeat(ExamRoom* obj, int value) {
    int pos = findInsertPosition(obj, value);
    if (pos < obj->size && obj->seats[pos] == value) {
        for (int i = pos; i < obj->size - 1; ++i) {
            obj->seats[i] = obj->seats[i + 1];
        }
        obj->size--;
    }
}

int examRoomSeat(ExamRoom* obj) {
    if (obj->size == 0) {
        insertSeat(obj, 0);
        return 0;
    }

    int left = obj->seats[0];
    int diff = left;
    int max = 0;

    // 중간 위치 계산
    for (int i = 1; i < obj->size; i++) {
        int right = obj->seats[i];
        int mid = left + (right - left) / 2;
        int distance = mid - left < right - mid ? mid - left : right - mid;
        if (distance > diff) {
            diff = distance;
            max = mid;
        }
        left = right;
    }

    // 마지막 자리와의 거리 확인
    int last = obj->seats[obj->size - 1];
    if (obj->n - last > diff) {
        max = obj->n;
    }

    // 선택된 좌석 추가
    insertSeat(obj, max);
    return max;
}

void examRoomLeave(ExamRoom* obj, int p) {
    removeSeat(obj, p);
}

void examRoomFree(ExamRoom* obj) {
    free(obj->seats);
    free(obj);
}

// 테스트용 main 함수
int main() {
    ExamRoom* obj = examRoomCreate(8);
    printf("Seat: %d\n", examRoomSeat(obj)); // 첫 번째 학생이 앉을 좌석
    printf("Seat: %d\n", examRoomSeat(obj)); // 두 번째 학생이 앉을 좌석
    printf("Seat: %d\n", examRoomSeat(obj)); // 세 번째 학생이 앉을 좌석
    examRoomLeave(obj, 0);                   // 첫 번째 학생이 떠남
    examRoomLeave(obj, 7);                   // 마지막 학생이 떠남
    printf("Seat: %d\n", examRoomSeat(obj)); // 추가 학생이 앉을 좌석
    printf("Seat: %d\n", examRoomSeat(obj));
    printf("Seat: %d\n", examRoomSeat(obj));

    examRoomFree(obj); // 메모리 해제
    return 0;
}

File: week10/19/P4.py
class ExamRoom(object):
    def __init__(self, n):
        self.room = []
        self.interval = [n-1]
        self.n = n

    def seat(self):
        if (self.room == []):
            self.room.append(0)
            self.interval = [0] + self.interval
            return 0
        if (self.room[0] != 0):
            self.interval[0] *= 2
        if (self.room[-1] != self.n-1):
            self.interval[-1] *= 2
        maxlen = max(self.interval) // 2 * 2
        offset = 0
        for i in range(len(self.interval)):
            if (self.interval[i] >= maxlen):
                offset += maxlen//2
                if (i == 0 and self.room[0] != 0):
                    offset = 0
                break
            if (i == 0 and self.room[0] != 0):
                offset += self.interval[0]//2
            else:
                offset += self.interval[i]
        for j in range(len(self.room)):
            if (self.room[j] > offset and j == 0):
                self.room = [offset] + self.room
                self.interval = [0] + [self.interval[0]//2] + self.interval[1:]
            elif (self.room[j] < offset and j == len(self.room)-1):
                self.room.append(offset)
                self.interval = self.interval[:-1] + [self.interval[-1]//2] + [0]
            elif (self.room[j] < offset < self.room[j+1]):
                self.room = self.room[:j+1] + [offset] + self.room[j+1:]
                self.interval = self.interval[:j+1] + [maxlen//2] + [self.interval[j+1]-maxlen//2] + self.interval[j+2:]
        if (self.room[0] != 0):
            self.interval[0] //= 2
        if (self.room[-1] != self.n-1):
            self.interval[-1] //= 2
        print(f"maxlen : {maxlen}, offset : {offset}")
        print(f"room : {self.room}, interval : {self.interval}")
        return offset

    def leave(self, p):
        i = self.room.index(p)
        self.room = self.room[:i] + self.room[i+1:]
        self.interval = self.interval[:i] + [self.interval[i]+self.interval[i+1]] + self.interval[i+2:]
        print(f"room : {self.room}, interval : {self.interval}")
        return

if __name__ == "__main__":
    obj = ExamRoom(8)
    obj.seat()
    obj.seat()
    obj.seat()
    obj.leave(0)
    obj.leave(7)
    obj.seat()
    obj.seat()
    obj.seat()
    obj.seat()
    obj.seat()
    obj.seat()
    obj.seat()